app/models/api/messages/useq_wafer_io.rb

```diff
@@ -0,0 +1,128 @@
+
+        def amp_assign_control_bead_tube
+          detect_descriptor('Assign Control Bead Tube')
+        end
+
+        def amp_instrument_name
+          detect_descriptor('UG AMP Inst. Name')
+        end
+      end
+    end
+  end
+
+  # Batch is the main / top-level model that contributes to the message sent to the MLWH.
+  renders_model(::Batch)
+
+  # The following section maps attributes on Sequencescape models to attributes in the json message that is passed to
+  # the MLWH.
+  # The following methods come from the Batch model or the relevant module above.
+  map_attribute_to_json_attribute(:id, 'wafer_id')
+
+  map_attribute_to_json_attribute(:updated_at)
+
+  # The following methods come from the Request model or the relevant module above.
+  # They are included in the MLWH message under 'lanes'.
+  with_nested_has_many_association(:requests, as: :lanes) do
+    map_attribute_to_json_attribute(:position, 'lane')
+    map_attribute_to_json_attribute(:mx_library, 'id_pool_lims')
+    map_attribute_to_json_attribute(:lane_identifier, 'entity_id_lims')
+    map_attribute_to_json_attribute(:otr_instrument_name)
+    map_attribute_to_json_attribute(:amp_instrument_name)
+    map_attribute_to_json_attribute(:otr_carrier_lot_number)
+    map_attribute_to_json_attribute(:otr_carrier_expiry)
+    map_attribute_to_json_attribute(:otr_reaction_mix_7_lot_number)
+    map_attribute_to_json_attribute(:otr_reaction_mix_7_expiry)
+    map_attribute_to_json_attribute(:otr_nfw_lot_number)
+    map_attribute_to_json_attribute(:otr_nfw_expiry)
+    map_attribute_to_json_attribute(:otr_oil_lot_number)
+    map_attribute_to_json_attribute(:otr_oil_expiry)
+    map_attribute_to_json_attribute(:otr_pipette_carousel)
+    map_attribute_to_json_attribute(:amp_assign_control_bead_tube)
+    map_attribute_to_json_attribute(:ot_recipe)
+
+    # The following methods come from the Aliquot model or the relevant module above.
+    # They are included in the MLWH message under 'samples'.
+    with_nested_has_many_association(:lane_samples, as: :samples) do
+      with_association(:tag) do
+        map_attribute_to_json_attribute(:oligo, 'tag_sequence')
+      end
+      map_attribute_to_json_attribute(:library_type, 'pipeline_id_lims')
+      with_association(:bait_library) { map_attribute_to_json_attribute(:name, 'bait_name') }
+      map_attribute_to_json_attribute(:insert_size_from, 'requested_insert_size_from')
+      map_attribute_to_json_attribute(:insert_size_to, 'requested_insert_size_to')
+      with_association(:sample) { map_attribute_to_json_attribute(:uuid, 'sample_uuid') }
+      with_association(:study) { map_attribute_to_json_attribute(:uuid, 'study_uuid') }
+      with_association(:primer_panel) { map_attribute_to_json_attribute(:name, 'primer_panel') }
+      with_association(:library) { map_attribute_to_json_attribute(:external_identifier, 'id_library_lims') }
+      map_attribute_to_json_attribute(:aliquot_type, 'entity_type')
+    end
```

Comment:
question: do we have control samples with this pipeline?

Reply:
My understanding is 'controls' in the warehouse are phix controls provided at batch creation (for aviti and illumina). Since ultima doesnt use phix or add external controls (to my knowledge) we don't have anything to send.